src/java.base/share/classes/jdk/internal/lang/StableValue.java

```diff
@@ -0,0 +1,386 @@
+ * <p>
+ * A stable value is said to be monotonic because the state of a stable value can only go
+ * from <em>unset</em> to <em>set</em> and consequently, a value can only be set
+ * at most once.
+ <p>
+ * To create a new fresh (unset) StableValue, use the {@linkplain StableValue#of()}
+ * factory.
+ * <p>
+ * To create collections of <em>wrapped stable elements</em>, that, in turn, are also
+ * eligible for constant folding optimizations, the following factories can be used:
+ * <ul>
+ *     <li>{@linkplain StableValue#ofList(int)}</li>
+ *     <li>{@linkplain StableValue#ofMap(Set)}</li>
+ *</ul>
+ * <p>
+ * Except for a StableValue's value itself, all method parameters must be <em>non-null</em>
+ * and all collections provided must only contain <em>non-null</em> elements or a
+ * {@link NullPointerException} will be thrown.
+ *
+ * @param <V> value type
+ * @since 23
+ */
+public sealed interface StableValue<V>
+        permits StableValueImpl,
+        StableValueElement {
+
+    /**
+     * {@return the set value (nullable) if set, otherwise throws
+     * {@code NoSuchElementException}}
+     *
+     * @throws NoSuchElementException if no value is set
+     */
+    V orThrow();
+
+    /**
+     * {@return {@code true} if a value is set, otherwise {@code false}}
+     */
+    boolean isSet();
+
+    /**
+     * Sets the stable value to the provided (nullable) {@code value} or throws an
+     * {@linkplain IllegalStateException} if a value is already set.
+     *
+     * @param value to set (nullable)
+     * @throws IllegalStateException if a value is already set
+     */
+    void setOrThrow(V value);
+
+    /**
+     * If no value is set, sets the stable value to the provided (nullable)
+     * {@code value}, returning the (pre-existing or newly set) value.
+     * <p>
+     * If several threads invoke this method simultaneously, only one thread will succeed
+     * in setting a value and that (witness) value will be returned to all threads.
+     *
+     * @param value to set (nullable)
+     * @return the bound value
+     */
+    V setIfUnset(V value);
+
+    /**
+     * If no value {@linkplain #isSet() is set}, attempts to compute and set a
+     * new (nullable) value using the provided {@code supplier}, returning the
+     * (pre-existing or newly set) value.
+     *
+     * <p>
+     * If the supplier throws an (unchecked) exception, the exception is rethrown, and no
+     * value is set. The most common usage is to construct a new object serving as a
+     * lazily computed value or memoized result, as in:
+     *
+     * <pre> {@code
+     * Value witness = stable.computeIfUnset(Value::new);
+     * }</pre>
+     *
+     * @implSpec The implementation logic is equivalent to the following steps for this
+     * {@code stable}:
+     *
+     * <pre> {@code
+     * if (stable.isBound()) {
+     *     return stable.get();
+     * } else {
+     *     V newValue = supplier.get();
+     *     stable.setOrThrow(newValue);
```

Comment:
If ::computeIfUnset allows racy sets, then it isn't equivalent to this code as ::setOrThrow will throw on a race, correct?

Reply:
`computeIfUnset()` is indeed guarded by a synchronized block, only it sits on the method declaration of `computeIfUnsetVolatile0`. I think we should have an internal mutex. This will also correspond to the stable collections which have internal mutexes for each index/key.